ros/src/computing/planning/motion/packages/astar_planner/nodes/velocity_set/libvelocity_set.h

```diff
@@ -233,13 +240,59 @@ class ObstacleTracker
 
   double calcVelocity(const geometry_msgs::Point& cpos)
   {
-    tf::Vector3 cx;
-    tf::pointMsgToTF(cpos, cx);
-    double dx = (position_buf_[1]-cx).length()-(position_buf_[0]-cx).length();
-    ros::Duration dt = time_buf_[1]-time_buf_[0];
-    kf_.predict();
-    double v = kf_.update(dx/dt.toSec());
-    v = (v > moving_thres_) ? v : 0.0;
+    double v;
+    ros::Duration dt = time_buf_[1]-time_buf_[0]; double dx;
+    // dt can be 0, when this function is called with out calling update()
+    // if dt is non zero
+    if (dt != ros::Duration(0.0)) {
+
+        // compute dx
+        tf::Vector3 cx;
+        tf::pointMsgToTF(cpos, cx);
+        dx = (position_buf_[1] - cx).length() - (position_buf_[0] - cx).length();
+
+        // dx can be negative for couple of frames, if a car cuts in when we were already tracking.
+        // if dx is non negative
+        if (dx >= 0) {
+            // KF estimate
+            kf_.predict();
+            v = kf_.update(dx / dt.toSec());
+            v = (v > moving_thres_) ? v : 0.0;
+
+            // update previous velocity
+            prev_velocity_ = v;
+            std::cerr << "prev vel: " << prev_velocity_ << "\n";
+
+            // reset the count
+            dx_negative_count_ = 0;
+        }
+        else
+        {
+            dx_negative_count_++;
+            std::cerr << "dx negative\n";
+
+            // then the car is moving towards us
+            if (dx_negative_count_ >= OBJECT_MOVING_TOWARD_US_THRESHOLD)
+            {
+                // zero velocity
+                v = 0.0;
+
+                //wrap the counter
+                dx_negative_count_ = OBJECT_MOVING_TOWARD_US_THRESHOLD;
+            }
+            else
+            {
+                v = prev_velocity_;
```

Comment:
in this case we just keep going with whatever the v was before we started getting negative dx?

Reply:
correct, we keep the prev_velocity_